srcs/dfa.cc

```diff
@@ -1,109 +1,46 @@
 #include <cassert>
-#include <vector>
-#include <unordered_set>
+
 #include "dfa.hpp"
 #include "token.hpp"
 
-std::unique_ptr<Dfa> Dfa::BuildDfa(){
-    Builder b{};
-    return b.AddStates().LinkStates().SetAcceptStates().SetStartState().Build();
-}
-
-
-void Dfa::Reset(){
-    _currentState = _startState;
-    _passedAcceptState = nullptr;
-    _passedAcceptStateStr.clear();
-    _buffer.clear();
+Dfa::State Dfa::AddState(Dfa::Action action, bool accepting){
```

Comment:
You can make things like ImmutableDFA and MutableDFA (or DFABuiler if you like), and only the latter has non-const methods.  The member variables will be protected instead of private.

To me, not much difference since you can disable the mutators by passing a const reference.